cartridge/roles/coordinator.lua

```diff
@@ -123,68 +108,76 @@ local function control_loop(conn)
             end
         end
 
+        local now = fiber.time()
         if next(updates) ~= nil then
-            local ok, err = errors.netbox_call(conn, 'set_leaders',
-                {updates}, {timeout = vars.options.NETBOX_CALL_TIMEOUT}
-            )
+            local ok, err = session:set_leaders(updates)
             if ok == nil then
                 log.error('%s', err)
                 break
             end
         end
 
-        local now = fiber.time()
         local next_moment = math.huge
         for _, decision in pairs(ctx.decisions) do
-            if (now < decision.immunity)
+            if (decision.immunity >= now)
             and (decision.immunity < next_moment)
             then
                 next_moment = decision.immunity
             end
         end
 
+        assert(next_moment >= now)
         vars.membership_notification:wait(next_moment - now)
-    until not pcall(fiber.testcancel)
+    until not pcall(fiber.testcancel) or not session:is_locked()
```

Comment:
Looks strange.

`until not pcall(fiber.testcancel)` is an infinite loop. not session:is_locked() potentially is not infinite. And this mixture looks strange.

Consider a way to move `not session:is_locked()` into loop body and add explicit break.